cmd/bucket-metadata.go

```diff
@@ -0,0 +1,136 @@
+/*
+ * Minio Cloud Storage, (C) 2014-2016 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"encoding/json"
+	"net/rpc"
+)
+
+// BucketMetaState - Interface to update bucket metadata in-memory
+// state.
+type BucketMetaState interface {
+	// Updates bucket notification
+	UpdateBucketNotification(args *SetBNPArgs) error
+
+	// Updates bucket listener
+	UpdateBucketListener(args *SetBLPArgs) error
+
+	// Updates bucket policy
+	UpdateBucketPolicy(args *SetBPPArgs) error
+
+	// Sends event
+	SendEvent(args *EventArgs) error
+}
+
+// Type that implements BucketMetaState for local node.
+type localBMS struct {
+	ObjectAPI func() ObjectLayer
+}
+
+func (lc *localBMS) UpdateBucketNotification(args *SetBNPArgs) error {
+	// check if object layer is available.
+	objAPI := lc.ObjectAPI()
+	if objAPI == nil {
+		return errServerNotInitialized
+	}
+
+	globalEventNotifier.SetBucketNotificationConfig(args.Bucket, args.NCfg)
+
+	return nil
+}
+
+func (lc *localBMS) UpdateBucketListener(args *SetBLPArgs) error {
+	// check if object layer is available.
+	objAPI := lc.ObjectAPI()
+	if objAPI == nil {
+		return errServerNotInitialized
+	}
+
+	// Update in-memory notification config.
+	return globalEventNotifier.SetBucketListenerConfig(args.Bucket, args.LCfg)
+}
+
+func (lc *localBMS) UpdateBucketPolicy(args *SetBPPArgs) error {
+	// check if object layer is available.
+	objAPI := lc.ObjectAPI()
+	if objAPI == nil {
+		return errServerNotInitialized
+	}
+
+	var pCh policyChange
+	if err := json.Unmarshal(args.PChBytes, &pCh); err != nil {
+		return err
+	}
+
+	return globalBucketPolicies.SetBucketPolicy(args.Bucket, pCh)
+}
+
+func (lc *localBMS) SendEvent(args *EventArgs) error {
+	// check if object layer is available.
+	objAPI := lc.ObjectAPI()
+	if objAPI == nil {
+		return errServerNotInitialized
+	}
+
+	return globalEventNotifier.SendListenerEvent(args.Arn, args.Event)
+}
+
+// Type that implements BucketMetaState for remote node.
+type remoteBMS struct {
+	*AuthRPCClient
+}
+
+func (rc *remoteBMS) UpdateBucketNotification(args *SetBNPArgs) error {
+	reply := GenericReply{}
+	err := rc.Call("S3.SetBucketNotificationPeer", args, &reply)
+	// Check for network error and retry once.
+	if err != nil && err.Error() == rpc.ErrShutdown.Error() {
```

Comment:
@harshavardhana We had a discussion about this earlier regarding this line https://github.com/minio/minio/blob/master/cmd/net-rpc-client.go#L145 - I had proposed changing it to `if err == rpc.ErrShutdown` - my understanding was that the way we could lose/hide the type was by using things like `traceError()`, otherwise this change would work. We do not at this time use `traceError()` here (in net-rpc-client.go) since the lowerlevel code is directly in the net/rpc package. It's just that we do `traceError()` with somewhat certain goals, and we don't do it uniformly, so comparing `.Error()` values is more reliable, in case future code changed underlying error types.

Perhaps we could consider this aspect in a separate issue and move ahead with this change?